File: domain/usecases/search_packages.py
from typing import List
from domain.entities import Package
from domain.repositories import PackageRepository

class SearchPackagesUseCase:
    def __init__(self, alpm_repo: PackageRepository, aur_repo: PackageRepository, flatpak_repo: PackageRepository):
        self.alpm_repo = alpm_repo
        self.aur_repo = aur_repo
        self.flatpak_repo = flatpak_repo

    def execute(self, query: str) -> List[Package]:
        if not query:
            return []
            
        results = []
        
        # 1. Search Pacman
        alpm_results = self.alpm_repo.search(query)
        results.extend(alpm_results)
        
        pacman_names = {pkg.name for pkg in alpm_results}
        
        # 2. Search AUR
        aur_results = self.aur_repo.search(query)
        # Filter duplicates where name matches pacman packages
        aur_filtered = [pkg for pkg in aur_results if pkg.name not in pacman_names]
        results.extend(aur_filtered)
        
        # 3. Search Flatpak
        flatpak_results = self.flatpak_repo.search(query)
        results.extend(flatpak_results)
        
        return results

File: src-rust/src/domain/usecases/search_packages.rs
use std::sync::Arc;
use crate::domain::{entities::Package, repositories::PackageRepository};
use crate::config::AppConfig;

pub struct SearchPackagesUseCase {
    pub alpm_repo: Arc<dyn PackageRepository>,
    pub aur_repo: Arc<dyn PackageRepository>,
    pub flatpak_repo: Arc<dyn PackageRepository>,
}

impl SearchPackagesUseCase {
    pub fn new(
        alpm_repo: Arc<dyn PackageRepository>,
        aur_repo: Arc<dyn PackageRepository>,
        flatpak_repo: Arc<dyn PackageRepository>,
    ) -> Self {
        Self { alpm_repo, aur_repo, flatpak_repo }
    }

    pub async fn execute(&self, query: &str, config: &AppConfig) -> Vec<Package> {
        if query.trim().len() < 3 {
            return vec![];
        }

        let alpm_fut = self.alpm_repo.search(query);
        let aur_fut = async {
            if config.enable_aur {
                self.aur_repo.search(query).await
            } else {
                vec![]
            }
        };
        let flatpak_fut = async {
            if config.enable_flatpak {
                self.flatpak_repo.search(query).await
            } else {
                vec![]
            }
        };

        let (alpm, aur, flatpak) = tokio::join!(alpm_fut, aur_fut, flatpak_fut);

        let mut results = alpm;
        results.extend(aur);
        results.extend(flatpak);
        results
    }
}
